### PR TITLE
Remove DistributedMeshTools.cpp and update mesh refinement

### PR DESCRIPTION
**Pull request by chrisrichardson**
* Removes `DistributedMeshTools::`
* Updates and improves mesh refinement code - rather than adding new vertices at the end of the global range, now expands the local indexing, and adds locally, avoiding a global sort.
* Various minor improvements to mesh refinement, making more modular. More can be done in a further PR.
* Some improvements to `graph::Partitioning::distribute_data` to allow variable size on input, and distribute correctly.

### CODE REVIEW

File: cpp/dolfinx/graph/Partitioning.h [hunk 1]
```diff
@@ -162,77 +162,31 @@ Partitioning::distribute_data(
...
-  Eigen::Array<T, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> x_recv(
-      disp_recv.back(), gdim);
-  MPI_Alltoallv(x_send.data(), number_send.data(), disp_send.data(),
-                compound_type, x_recv.data(), number_recv.data(),
-                disp_recv.data(), compound_type, comm);
+  const int rank = dolfinx::MPI::rank(comm);
+  std::vector<std::int64_t> global_offsets(size + 1, 0);
+  MPI_Allgather(&num_points_local, 1, MPI_INT64_T, global_offsets.data() + 1, 1,
+                MPI_INT64_T, comm);
+  for (int i = 0; i < size; ++i)
```
Comment (garth-wells) [inline cpp/dolfinx/graph/Partitioning.h:171]: Use `std::partial_sum` here.

File: cpp/dolfinx/graph/Partitioning.h [hunk 2]
```diff
@@ -275,12 +230,16 @@ Partitioning::distribute_data(
     for (int i = disp_index_recv[p]; i < disp_index_recv[p + 1]; ++i)
     {
       const std::int64_t index = indices_recv[i];
```
Comment (garth-wells) [inline cpp/dolfinx/graph/Partitioning.h:232]: Remove this temporary for extra clarity.

File: cpp/dolfinx/graph/Partitioning.h [hunk 3]
```diff
@@ -275,12 +230,16 @@ Partitioning::distribute_data(
...
-      const std::int32_t index_local = index - offset_x;
+      const std::int32_t index_local = index - global_offsets[rank];
       assert(index_local >= 0);
-      x_return.row(i) = x_recv.row(index_local);
+      x_return.row(i) = x.row(index_local);
     }
   }
 
+  MPI_Datatype compound_type;
+  MPI_Type_contiguous(gdim, dolfinx::MPI::mpi_type<T>(), &compound_type);
```
Comment (garth-wells) [inline cpp/dolfinx/graph/Partitioning.h:240]: `gdim` -> `item_size`?

File: cpp/dolfinx/refinement/ParallelRefinement.cpp [hunk 4]
```diff
@@ -23,6 +22,66 @@
 using namespace dolfinx;
 using namespace dolfinx::refinement;
 
+namespace
+{
+
+//-----------------------------------------------------------------------------
+// Create geometric points of new Mesh, from current Mesh and a edge_to_vertex
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.cpp:29]: Use doxygen formatting.

File: cpp/dolfinx/refinement/ParallelRefinement.cpp [hunk 5]
```diff
@@ -160,90 +213,43 @@ void ParallelRefinement::update_logical_edgefunction()
...
+  _new_vertex_coordinates
+      = create_new_geometry(_mesh, local_edge_to_new_vertex);
 
   // If they are shared, then the new global vertex index needs to be
-  // sent off-process.  Add offset to map, and collect up any shared new
-  // vertices that need to send the new index off-process
+  // sent off-process.
 
   int num_neighbours = _marked_for_update.size();
   std::vector<std::vector<std::int64_t>> values_to_send(num_neighbours);
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.cpp:250]: Can we avoid `std::vector<std::vector>>`?

File: cpp/dolfinx/refinement/ParallelRefinement.cpp [hunk 6]
```diff
@@ -282,33 +288,54 @@ void ParallelRefinement::create_new_vertices()
...
-
-  _new_vertex_coordinates
-      = mesh::DistributedMeshTools::reorder_by_global_indices(
-          _mesh.mpi_comm(), _new_vertex_coordinates, global_indices);
+  return local_edge_to_new_vertex;
 }
 //-----------------------------------------------------------------------------
-mesh::Mesh ParallelRefinement::build_local() const
+std::vector<std::int64_t>
+ParallelRefinement::adjust_indices(std::shared_ptr<const common::IndexMap> im,
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.cpp:300]: Use more informative name than `im`.

Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.cpp:300]:
```cpp
const std::shared_ptr<const common::IndexMap>&
```

File: cpp/dolfinx/refinement/ParallelRefinement.cpp [hunk 7]
```diff
@@ -282,33 +288,54 @@ void ParallelRefinement::create_new_vertices()
...
+  // and adjust existing indices to match.
+
+  // Get number of new indices on all processes
+  int mpi_size = dolfinx::MPI::size(im->mpi_comm());
+  int mpi_rank = dolfinx::MPI::rank(im->mpi_comm());
+  std::vector<std::int32_t> recvn(mpi_size);
+  MPI_Allgather(&n, 1, MPI_INT32_T, recvn.data(), 1, MPI_INT32_T,
+                im->mpi_comm());
+  std::vector<std::int64_t> global_offsets = {0};
+  for (std::int32_t r : recvn)
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.cpp:313]: Use `reserve` on `global_offsets`

File: cpp/dolfinx/refinement/ParallelRefinement.cpp [hunk 8]
```diff
@@ -282,33 +288,54 @@ void ParallelRefinement::create_new_vertices()
...
+  for (int i = 0; i < local_size; ++i)
+    global_indices[i] += global_offsets[mpi_rank];
+  for (int i = 0; i < ghost_owners.size(); ++i)
+    global_indices[local_size + i] += global_offsets[ghost_owners[i]];
+
+  return global_indices;
+}
+//-----------------------------------------------------------------------------
+mesh::Mesh
+ParallelRefinement::build_local(std::vector<std::int64_t>& cell_topology) const
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.cpp:329]: Add `const`,
```cpp
const std::vector<std::int64_t>&
```

File: cpp/dolfinx/refinement/ParallelRefinement.cpp [hunk 9]
```diff
@@ -317,13 +344,15 @@ mesh::Mesh ParallelRefinement::build_local() const
   return mesh;
 }
 //-----------------------------------------------------------------------------
-mesh::Mesh ParallelRefinement::partition(bool redistribute) const
+mesh::Mesh
+ParallelRefinement::partition(std::vector<std::int64_t>& cell_topology,
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.cpp:348]: Add `const`

File: cpp/dolfinx/refinement/ParallelRefinement.h [hunk 10]
```diff
@@ -70,40 +75,39 @@ class ParallelRefinement
...
-  const std::map<std::int32_t, std::int64_t>& edge_to_new_vertex() const;
-
-  /// Add new cells with vertex indices
-  /// @param[in] idx
-  void new_cells(const std::vector<std::int64_t>& idx);
+  /// @return edge_to_new_vertex map
+  std::map<std::int32_t, std::int64_t> create_new_vertices();
 
   /// Use vertex and topology data to partition new mesh across processes
+  /// @param[in] cell_topology
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.h:82]: Add explanation of the two arguments

File: cpp/dolfinx/refinement/ParallelRefinement.h [hunk 11]
```diff
@@ -70,40 +75,39 @@ class ParallelRefinement
...
   /// @return A Mesh
-  mesh::Mesh build_local() const;
+  mesh::Mesh build_local(std::vector<std::int64_t>& cell_topology) const;
+
+  /// Adjust indices to account for extra n values on each process
+  /// This is a utility to help add new topological vertices on each process
+  /// into the space of the index map.
+  ///
+  /// @param im IndexMap
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.h:97]: Add explanation

File: cpp/dolfinx/refinement/ParallelRefinement.h [hunk 12]
```diff
@@ -70,40 +75,39 @@ class ParallelRefinement
...
+  /// @param n Number of new entries to be accommodated on this process
+  /// @return Global indices as if "n" extra values are appended on each process
+  static std::vector<std::int64_t>
+  adjust_indices(std::shared_ptr<const common::IndexMap> im, std::int32_t n);
 
 private:
   // mesh::Mesh reference
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.h:104]: Remove redundant comment.

File: cpp/dolfinx/refinement/ParallelRefinement.h [hunk 13]
```diff
@@ -70,40 +75,39 @@ class ParallelRefinement
...
-  std::map<std::int32_t, std::int64_t> _local_edge_to_new_vertex;
-
   // New storage for all coordinates when creating new vertices
   Eigen::Array<double, Eigen::Dynamic, 3, Eigen::RowMajor>
       _new_vertex_coordinates;
 
-  // New storage for all cells when creating new topology
-  std::vector<std::int64_t> _new_cell_topology;
-
   // Management of marked edges
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.h:111]: Make comment meaningful

File: cpp/dolfinx/refinement/ParallelRefinement.h [hunk 14]
```diff
@@ -70,40 +75,39 @@ class ParallelRefinement
...
 
 private:
   // mesh::Mesh reference
   const mesh::Mesh& _mesh;
 
-  // Mapping from old local edge index to new global vertex, needed to
-  // create new topology
-  std::map<std::int32_t, std::int64_t> _local_edge_to_new_vertex;
-
   // New storage for all coordinates when creating new vertices
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/ParallelRefinement.h:107]: Why is the stored?

File: cpp/dolfinx/refinement/PlazaRefinementND.cpp [hunk 15]
```diff
@@ -123,10 +127,8 @@ mesh::Mesh compute_refinement(const mesh::Mesh& mesh, ParallelRefinement& p_ref,
     if (marked_edge_list.size() == 0)
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/PlazaRefinementND.cpp:127]: `if (marked_edge_list.empty())

File: cpp/dolfinx/refinement/PlazaRefinementND.cpp [hunk 16]
```diff
@@ -175,18 +177,16 @@ mesh::Mesh compute_refinement(const mesh::Mesh& mesh, ParallelRefinement& p_ref,
         parent_cell.push_back(c);
 
       // Convert from cell local index to mesh index and add to cells
-      std::vector<std::int64_t> simplex_set_global(simplex_set.size());
-      for (std::size_t i = 0; i < simplex_set_global.size(); ++i)
-        simplex_set_global[i] = indices[simplex_set[i]];
-      p_ref.new_cells(simplex_set_global);
+      for (std::size_t i = 0; i < simplex_set.size(); ++i)
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/PlazaRefinementND.cpp:180]: Use range-based loop.

File: cpp/dolfinx/refinement/PlazaRefinementND.cpp [hunk 17]
```diff
@@ -175,18 +177,16 @@ mesh::Mesh compute_refinement(const mesh::Mesh& mesh, ParallelRefinement& p_ref,
...
+        cell_topology.push_back(indices[simplex_set[i]]);
     }
   }
 
   const bool serial = (dolfinx::MPI::size(mesh.mpi_comm()) == 1);
```
Comment (garth-wells) [inline cpp/dolfinx/refinement/PlazaRefinementND.cpp:185]: Just use MPI size.

File: cpp/dolfinx/refinement/ParallelRefinement.h (hunk 13 above)
Comment (chrisrichardson) [reply cpp/dolfinx/refinement/ParallelRefinement.h:111]: This code has been removed

File: cpp/dolfinx/refinement/ParallelRefinement.h (hunk 14 above)
Comment (chrisrichardson) [reply cpp/dolfinx/refinement/ParallelRefinement.h:107]: This code has been removed

File: cpp/dolfinx/refinement/ParallelRefinement.cpp (hunk 5 above)
Comment (chrisrichardson) [reply cpp/dolfinx/refinement/ParallelRefinement.cpp:250]: Possible by traversing the data twice, but makes the code longer and more complex, and I am not sure it is more efficient. The number of vectors here is quite limited, as it is `num_neighbours` not `mpi_size`.

**Pull request merged**